cnd/src/network/protocol.rs

```diff
@@ -0,0 +1,207 @@
+    inner: T,
+}
+
+impl<T> fmt::Debug for PushSubstream<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("PushSubstream").finish()
+    }
+}
+
+impl<T> PushSubstream<T>
+where
+    T: AsyncWrite + Unpin,
+{
+    /// Sends the information we wish to push on the substream.
+    ///
+    /// Consumes the substream, returning a future that resolves
+    /// when the info has been sent on the underlying connection.
+    pub async fn send(mut self, info: SecretHashInfo) -> Result<(), io::Error> {
+        debug!("Sending secret hash info to peer");
+
+        let message = serde_json::to_vec(&info).unwrap();
+        upgrade::write_one(&mut self.inner, &message).await
+    }
+}
+
+impl UpgradeInfo for SecretHashProtocolConfig {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/comit/swap/secret_hash/1.0.0")
+    }
+}
+
+impl<C> InboundUpgrade<C> for SecretHashProtocolConfig
+where
+    C: AsyncRead + AsyncWrite + Unpin + Send + Sync + 'static,
+{
+    type Output = SecretHashInfo;
+    type Error = upgrade::ReadOneError;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn upgrade_inbound(self, mut socket: C, _: Self::Info) -> Self::Future {
+        trace!("Upgrading inbound connection");
+        Box::pin(async move {
+            let msg = upgrade::read_one(&mut socket, *JSON_INFO_LENGTH).await?;
+            let mut de = serde_json::Deserializer::from_slice(&msg);
+            let info = SecretHashInfo::deserialize(&mut de).unwrap();
+            Ok(info)
+        })
+    }
+}
+
+impl<C> OutboundUpgrade<C> for SecretHashProtocolConfig
+where
+    C: AsyncRead + AsyncWrite + Unpin + Send + Sync + 'static,
+{
+    type Output = PushSubstream<C>;
+    type Error = io::Error;
+    type Future = future::Ready<Result<Self::Output, Self::Error>>;
+
+    fn upgrade_outbound(self, socket: C, _: Self::Info) -> Self::Future {
+        trace!("Upgrading outbound connection");
+        future::ok(PushSubstream { inner: socket })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::swap_protocols::SwapId;
+    use futures::{channel::oneshot, prelude::*};
+    use libp2p::{
+        core::{
+            upgrade::{self, apply_inbound, apply_outbound},
+            Transport,
+        },
+        tcp::TcpConfig,
+    };
+
+    static SWAP_ID: SwapId = SwapId::from_str("7f3a105d-ecf2-4cc6-b35c-b4351ac28a34").unwrap();
+
+    #[test]
+    fn swap_id_format_has_not_changed() {
+        let static_swap_id = SWAP_ID;
+        let default_swap_id = SwapId::default();
+
+        let static_swap_id = format!("{}", static_swap_id);
+        let default_swap_id = format!("{}", default_swap_id);
+
+        let want = static_swap_id.len();
+        let got = default_swap_id.len();
+
+        assert_eq!(want, got)
+    }
+
+    #[test]
+    fn correct_transfer() {
```

Comment:
I think your test will abort too early because you are using spawn two times ...

You need to block/await on something.

Reply:
We are going to write several of these tests. I suggest to refactor something reusable out of this as soon as we write the 2nd one to not duplicate too much.